torchdata/datapipes/iter/util/cacheholder.py

```diff
@@ -223,16 +232,20 @@ def _cache_check_fn(data, filepath_fn, hash_dict, hash_type, extra_check_fn):
                 with portalocker.Lock(promise_filepath, "a+", flags=portalocker.LockFlags.EXCLUSIVE) as promise_fh:
                     promise_fh.seek(0)
                     data = promise_fh.read()
-                    # TODO(VitalyFedyunin): Potentially there is old .promise file from previous failed run, we
-                    # need to somehow propagate uniq session id for dataloader, save and compare it here,
-                    # raising error
+                    _check_promise_file_uuid(data, promise_filepath, filepath, uuid)
                     file_exists = len(data) > 0
                     if not file_exists:
                         result = False
                         promise_fh.seek(0)
-                        promise_fh.write("[dataloader session uid]")
+                        promise_fh.write(str(uuid))
                         promise_fh.truncate()
                         promise_fh.flush()
+            else:
+                promise_filepath = _promise_filename(filepath)
+                with portalocker.Lock(promise_filepath, "r", flags=portalocker.LockFlags.SHARED) as promise_fh:
```

Comment:
IIUC, there can be a case that `promise` file has been removed when the promise if fulfilled.